File: Iteration2/src/Domain/Grade.java
package Domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
@JsonIgnoreProperties(ignoreUnknown = true)
public class Grade {
    private  Course course;
    private  float numGrade;
    private String letterGrade;
    private boolean isPassed;
    private float numericGradeForGano;

    public Grade(Course course, float numericGrade) {
        this.course = course;
        this.numGrade = numericGrade;
        letterGradeCalculation();
    }
    public Grade(){

    }
    private void letterGradeCalculation() {
        if (numGrade >= 88) {
            isPassed = true;
            letterGrade = "AA";
            numericGradeForGano = 4;
        } else if (numGrade >= 81) {
            isPassed = true;
            letterGrade = "BA";
            numericGradeForGano = 3.5f;
        } else if (numGrade >= 74) {
            isPassed = true;
            letterGrade = "BB";
            numericGradeForGano = 3;
        } else if (numGrade >= 67) {
            isPassed = true;
            letterGrade = "CB";
            numericGradeForGano = 2.5f;
        } else if (numGrade >= 60) {
            isPassed = true;
            letterGrade = "CC";
            numericGradeForGano = 2;
        } else if (numGrade >= 53) {
            isPassed = true;
            letterGrade = "DC";
            numericGradeForGano = 1.5f;
        } else if (numGrade >= 46) {
            isPassed = true;
            letterGrade = "DD";
            numericGradeForGano = 1;
        }else {
            isPassed = false;
            letterGrade = "FF";
            numericGradeForGano = 0;
        }
    }

    @JsonIgnore
    public Course getCourse() {
        return course;
    }

    public float getNumericGrade() {
        return numGrade;
    }

    public String getLetterGrade() {
        return letterGrade;
    }


    public boolean isPassed() {
        return isPassed;
    }


    public void setPassed(boolean passed) {
        isPassed = passed;
    }
    public void setNumericGrade(float numericGrade){
        this.numGrade = numericGrade;
        letterGradeCalculation();
    }

    public float getNumericGradeForGano() {
        return numericGradeForGano;
    }


}

File: Iteration3/Domain/Grade.py
# Domain/Grade.py

class Grade:
    def __init__(self, course, numeric_grade):
        self.course = course
        self.num_grade = numeric_grade
        self._is_passed = None  # Initialize is_passed as None

        self.letter_grade_calculation()

    def letter_grade_calculation(self):
        if self.num_grade >= 88:
            self._is_passed = True
            self.letter_grade = "AA"
            self.numeric_grade_for_gano = 4
        elif self.num_grade >= 81:
            self._is_passed = True
            self.letter_grade = "BA"
            self.numeric_grade_for_gano = 3.5
        elif self.num_grade >= 74:
            self._is_passed = True
            self.letter_grade = "BB"
            self.numeric_grade_for_gano = 3
        elif self.num_grade >= 67:
            self._is_passed = True
            self.letter_grade = "CB"
            self.numeric_grade_for_gano = 2.5
        elif self.num_grade >= 60:
            self._is_passed = True
            self.letter_grade = "CC"
            self.numeric_grade_for_gano = 2
        elif self.num_grade >= 53:
            self._is_passed = True
            self.letter_grade = "DC"
            self.numeric_grade_for_gano = 1.5
        elif self.num_grade >= 46:
            self._is_passed = True
            self.letter_grade = "DD"
            self.numeric_grade_for_gano = 1
        else:
            self._is_passed = False
            self.letter_grade = "FF"
            self.numeric_grade_for_gano = 0


    def get_course(self):
        return self.course

    def get_numeric_grade(self):
        return self.num_grade

    def get_letter_grade(self):
        return self.letter_grade

    def is_passed(self):
        return self._is_passed

    def set_passed(self, passed):
        self._is_passed = passed

    def set_numeric_grade(self, numeric_grade):
        self.num_grade = numeric_grade
        self.letter_grade_calculation()

    def get_numeric_grade_for_gano(self):
        return self.numeric_grade_for_gano



import json

class GradeEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Grade):
            return {
                'course': obj.get_course(),
                'numeric_grade': obj.get_numeric_grade(),
                'letter_grade': obj.get_letter_grade(),
                'is_passed': obj.is_passed(),
                'numeric_grade_for_gano': obj.get_numeric_grade_for_gano()
            }
        return super().default(obj)
